[odh]Sync Upstream for v2.24

VaishnaviHire opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->

## Description
<!--- Describe your changes in detail -->

## How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->

## Merge criteria:
<!--- This PR will be merged by any repository approver when it meets all the points in the checklist -->
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->

- [ ] The commits are squashed in a cohesive manner and have meaningful messages.
- [ ] Testing instructions have been added in the PR body (for PRs involving changes that are not immediately obvious).
- [ ] The developer has manually tested the changes and verified that the changes work


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * TLS/CA bundle configuration for LlamaStackDistribution, including multi-key support and automatic pod restarts on bundle changes.
  * New viewer/editor RBAC roles for managing distributions.
* **Documentation**
  * Added CA bundle configuration guide; updated API overview and contributing/testing principles.
* **Samples**
  * New example manifests and a certificate generation script to demonstrate TLS setup.
* **Tests**
  * Added TLS end-to-end suite and strengthened e2e debugging and readiness checks.
* **Chores**
  * Introduced GitHub issue templates and config; added CPU-only image workflow; updated gitignore.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

leseb commented:
> You would need to update crd by doing `make generate manifests`

Thanks, I'm trying to figure out why it's failing locally:

```
make generate
/Users/leseb/go/src/github.com/opendatahub-io/llama-stack-operator/bin/controller-gen object:headerFile="hack/boilerplate.go.txt" paths="./..."
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
        panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x2 addr=0x0 pc=0x102e4b964]

goroutine 38 [running]:
go/types.(*Checker).handleBailout(0x14000583340, 0x1400122fd38)
        /opt/homebrew/Cellar/go/1.23.5/libexec/src/go/types/check.go:404 +0x9c
panic({0x10311db60?, 0x1036e0340?})
        /opt/homebrew/Cellar/go/1.23.5/libexec/src/runtime/panic.go:785 +0x124
go/types.(*StdSizes).Sizeof(0x0, {0x1031ee9a0, 0x1036e8e00})
        /opt/homebrew/Cellar/go/1.23.5/libexec/src/go/types/sizes.go:229 +0x314
go/types.(*Config).sizeof(...)
        /opt/homebrew/Cellar/go/1.23.5/libexec/src/go/types/sizes.go:334
go/types.representableConst.func1({0x1031ee9a0?, 0x1036e8e00?})
        /opt/homebrew/Cellar/go/1.23.5/libexec/src/go/types/const.go:77 +0x90
go/types.representableConst({0x1031f5238, 0x1036b4c60}, 0x14000583340, 0x1036e8e00, 0x0)
        /opt/homebrew/Cellar/go/1.23.5/libexec/src/go/types/const.go:93 +0x134
go/types.(*Checker).arrayLength(0x14000583340, {0x1031f35a8, 0x14001fd3820?})
        /opt/homebrew/Cellar/go/1.23.5/libexec/src/go/types/typexpr.go:537 +0x1bc
go/types.(*Checker).typInternal(0x14000583340, {0x1031f1bf8, 0x14001fd5950}, 0x0)
        /opt/homebrew/Cellar/go/1.23.5/libexec/src/go/types/typexpr.go:320 +0x3b0
go/types.(*Checker).definedType(0x14000583340, {0x1031f1bf8, 0x14001fd5950}, 0x1031efb48?)
        /opt/homebrew/Cellar/go/1.23.5/libexec/src/go/types/typexpr.go:201 +0x28
go/types.(*Checker).varType(0x14000583340, {0x1031f1bf8, 0x14001fd5950})
        /opt/homebrew/Cellar/go/1.23.5/libexec/src/go/types/typexpr.go:166 +0x2c
go/types.(*Checker).structType(0x14000583340, 0x14001fe9d70, 0x14001fe9d70?)
        /opt/homebrew/Cellar/go/1.23.5/libexec/src/go/types/struct.go:114 +0x11c
go/types.(*Checker).typInternal(0x14000583340, {0x1031f1b68, 0x14001fd0dc8}, 0x14001feeaf0)
        /opt/homebrew/Cellar/go/1.23.5/libexec/src/go/types/typexpr.go:337 +0xc20
go/types.(*Checker).definedType(0x14000583340, {0x1031f1b68, 0x14001fd0dc8}, 0x102f59d92?)
        /opt/homebrew/Cellar/go/1.23.5/libexec/src/go/types/typexpr.go:201 +0x28
go/types.(*Checker).typeDecl(0x14000583340, 0x14001feeaf0, 0x14001fd6d40, 0x0)
        /opt/homebrew/Cellar/go/1.23.5/libexec/src/go/types/decl.go:649 +0x454
go/types.(*Checker).objDecl(0x14000583340, {0x1031fa760, 0x14001feeaf0}, 0x0)
        /opt/homebrew/Cellar/go/1.23.5/libexec/src/go/types/decl.go:191 +0x874
go/types.(*Checker).packageObjects(0x14000583340)
        /opt/homebrew/Cellar/go/1.23.5/libexec/src/go/types/resolver.go:702 +0x350
go/types.(*Checker).checkFiles(0x14000583340, {0x14001772a68, 0x3, 0x3})
        /opt/homebrew/Cellar/go/1.23.5/libexec/src/go/types/check.go:459 +0x190
go/types.(*Checker).Files(0x14000188810?, {0x14001772a68?, 0x14001b43020?, 0x6?})
        /opt/homebrew/Cellar/go/1.23.5/libexec/src/go/types/check.go:422 +0x80
sigs.k8s.io/controller-tools/pkg/loader.(*loader).typeCheck(0x140003ca810, 0x1400020bd80)
        /Users/leseb/go/pkg/mod/sigs.k8s.io/controller-tools@v0.9.2/pkg/loader/loader.go:286 +0x2d8
sigs.k8s.io/controller-tools/pkg/loader.(*Package).NeedTypesInfo(0x1400020bd80)
        /Users/leseb/go/pkg/mod/sigs.k8s.io/controller-tools@v0.9.2/pkg/loader/loader.go:99 +0x44
sigs.k8s.io/controller-tools/pkg/loader.(*TypeChecker).check(0x14000e25110, 0x1400020bd80)
        /Users/leseb/go/pkg/mod/sigs.k8s.io/controller-tools@v0.9.2/pkg/loader/refs.go:268 +0x300
sigs.k8s.io/controller-tools/pkg/loader.(*TypeChecker).check.func1(0x37?)
        /Users/leseb/go/pkg/mod/sigs.k8s.io/controller-tools@v0.9.2/pkg/loader/refs.go:262 +0x54
created by sigs.k8s.io/controller-tools/pkg/loader.(*TypeChecker).check in goroutine 225
        /Users/leseb/go/pkg/mod/sigs.k8s.io/controller-tools@v0.9.2/pkg/loader/refs.go:260 +0x22c
make: *** [generate] Error 2
```

Any clue? Same issue on `main`.

leseb commented:
Waiting for https://github.com/opendatahub-io/llama-stack-k8s-operator/pull/26 :)

VaishnaviHire commented:
Yeah its a version mismatch with go. I has pinned the version to 1.21 in dockerfiles and go modules.

VaishnaviHire commented:
I think we can merge this and follow-up with CRD updates

nathan-weinberg committed:
chore: add github issue templates (#122)

similar to those in the llama-stack repo with some minor differences

Approved-by: mfleader
(cherry picked from commit 89e8f503288d6c73d0c3a8e8c7b419e93c018a3b)

mfleader committed:
feat(kustomizer): add service comparison and immutability enforcement utilities (#113)

Introduces a new `pkg/compare` package that provides utilities for detecting unexpected changes in Kubernetes Service resources by comparing desired vs current states and enforcing resource kind immutability by preventing users from mutating fields that should not be changed on existing resources. The package intelligently ignores operator-managed fields (ports, selectors) and cluster-managed metadata to focus only on detecting external mutations that could impact service functionality.

Approved-by: rhdedgar

Approved-by: VaishnaviHire
(cherry picked from commit f45c34a2cf1720be192f10c7042ee9d40458fbf7)

mfleader committed:
test(e2e): enhance debugging capabilities with detailed diagnostic explanations (#112)

Improved observability into system state during end-to-end tests by
logging state information from Distribution Status, Namespace Events,
Deployment, Pods, Services and Service Endpoints, and Operator Health.

---------

Signed-off-by: Matthew F Leader <mleader@redhat.com>
(cherry picked from commit 484e30ca1c1c431c22b4126ae066817fe68fd5cc)

mfleader committed:
refactor(controller): add anti-fragile test patterns and clean up utilities (#121)

Refactor controller tests to use `controllers_test` package with shared envtest environment. This consolidates test environment setup in `suite_test.go` using `TestMain` and adds anti-fragile testing patterns including a DistributionBuilder, consolidated test constants, and behavioral helpers. This pull request also documents the intent behind these changes in `CONTRIBUTING.md`.

This is intended to facilitate an upcoming pull request to add a `TestReconcile()` to the controller's test suite to more fully cover the controller's happy path to reconciliation.

Approved-by: VaishnaviHire
(cherry picked from commit 3b8639adde525b77ebe8eeb2652aa0cb98ed1366)

rhdedgar committed:
fix: vllm deployment script runAsNonRoot update (#126)

Setting `runAsNonRoot` to false for kubernetes clusters. The deployment issue currently happens on EKS and GKE clusters without this setting.

Closes #125

Approved-by: VaishnaviHire
(cherry picked from commit 5ce363513c065c08717d0aad15115c1b075cefe6)

mfleader committed:
test(controller): add integration test for reconciliation happy path (#123)

### Context
During [migration of the
Service](https://github.com/llamastack/llama-stack-k8s-operator/pull/105)
resource from Golang to a yaml manifest, I realized the controller's
`Reconcile()` needed increased test coverage to ensure the migration
code changes incurred no external behavior changes.

### Changes
- Tests Service, Service Account, Network Policy, and Deployment
coordination and port alignment
- Validates NetworkPolicy, Service, and Deployment creation
- Verifies owner references across all managed resources

---------

Signed-off-by: Matthew F Leader <mleader@redhat.com>
(cherry picked from commit 1710fa6ee6be895baf9d924c2fefd270b0c7fd8d)

mfleader committed:
feat(controller): migrate service reconciliation to kustomize manifest (#105)

Migrate service creation from programmatic to manifest-based Kustomizer pipeline, with added service comparison utilities to detect unexpected changes and enforce field mutability boundaries between operator-managed and cluster-managed state.

Approved-by: rhdedgar

Approved-by: VaishnaviHire
(cherry picked from commit 669f3e474bf0f42a3d215172e04a71c9e45e8316)

rhdedgar committed:
feat: support providers with self-signed certificates (#109)

- Updated the operator's CRD to allow the definition of a configMap containing ca-cert data, by configuring fields in a LlamaStackDistribution CR.
- Configured llama-stack to trust connections to providers like vLLM using server certificates from that Certificate Authority.
- Added an example script to generate generic self-signed certificates in PEM format to use with vLLM.
- Added an example configuration file to load the self-signed certs into a secret, mount the secret into the container, and point vLLM to the certs on server startup.

Approved-by: VaishnaviHire
(cherry picked from commit 780ac214293d7c92873e8c5a14c1d11e558e3a2f)

VaishnaviHire committed:
(fix): Preserve immutable fields during upgrade (#131)

During upgrades, if we do server-side apply without preserving the selectors field, we get 'field is immutable' error which blocks deployment updates and also causes continuous operator reconciliation. Preserving existing selector values fixes the issue.

Approved-by: rhdedgar

Approved-by: mfleader
(cherry picked from commit e2cc886ca0a49afa743ba7d7f2d19bcaefa846e6)

rhdedgar committed:
fix: updating CRD names; including llds editor/viewer roles (#137)

Adds a reference to the config/rbac/llsd_editor_role.yaml and config/rbac/llsd_viewer_role.yaml files to config/rbac/kustomization.yaml so the ClusterRoles get created upon installation.

Fixes: RHAIENG-496

Approved-by: leseb
(cherry picked from commit a7a964437eb4f2479c729268a6db9e42b5fc3ca3)

mfleader committed:
fix(plugins): handle nil intermediate values in JSON path traversal (#136)

Simplify field mutator implementation by leveraging jsonpointer v0.21.2 fix for nil intermediate values during field mutation. Replaces complex manual traversal with direct jsonpointer calls and streamlined path creation logic. Improved error handling through jsonpointer's improved error messages.

Previously worked around: panic: reflect: call of reflect.Value.Type on zero Value.

Approved-by: rhdedgar
(cherry picked from commit b11bd68d4b7e12370fe73d669b3114231268741d)

mfleader committed:
feat(controllers): enable testing of provider and version status via dependency injection (#133)

Refactors the LlamaStackDistributionReconciler to accept an httpClient
dependency. This makes the controller's status-update logic, which relies
on external API calls, testable in isolation.

The primary goal of this change is to enable a new integration test that
validates the controller correctly populates provider and version info
in the resource status. This test is added to formalize the expected
behavior and prevent regressions.

`checkHealth()` and its endpoint `v1/health` were not included in these changes since they will be removed in a subsequent code change, [130](https://github.com/llamastack/llama-stack-k8s-operator/pull/130).

Approved-by: rhdedgar
(cherry picked from commit 7cca6e45c8cd25d0f4b344cf22c55b88f114a254)

mfleader committed:
test(controler): add tests for handling network policy disablement (#140)

Adds behavioral test for handling disabled NetworkPolicy.

Approved-by: rhdedgar
(cherry picked from commit 3442b6bded012ba34631f7b41e9385da5f0e1544)

courtneypacheco committed:
feat(CI/CD): stub workflow for CPU-based vLLM image generation (#138)

Create a stub workflow to test PR #129.

Approved-by: rhdedgar
(cherry picked from commit 80df219c5f6b16473c102ac9f1376f4e8bb45332)

mfleader committed:
refactor(controller): delegate health checks to k8s readiness probes (#130)

The controller now relies on the Deployment's readiness status, which is driven by a new readiness probe on the container.

This removes the operator's redundant, manual health-checking logic and makes the system more robust by using a standard Kubernetes feature.

Approved-by: rhdedgar
(cherry picked from commit 2fa21ea25dc91524d86bfb69c053e16d12d5bd72)

rhdedgar reviewed (approved):
+1 here, I didn't notice any unexpected changes.
